src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs

```
@@ -0,0 +1,212 @@
+                        DataContentType = "application/json",
+                        Type = string.IsNullOrWhiteSpace(queueItem.Type) ? DefaultType : queueItem.Type,
+                        Source = new Uri(string.IsNullOrWhiteSpace(queueItem.Source) ? DefaultSource : queueItem.Source),
+                        Data = insertItem.TransformationResult
+                    };
+
+                    eventMessage.SetPartitionKey(insertItem.ChangeVector);
+
+                    events.Add(new KafkaMessageEvent
+                    {
+                        Topic = topicName, Message = eventMessage.ToKafkaMessage(ContentMode.Binary, jsonSerializer)
+                    });
+                }
+            }
+
+            if (events.Count == 0) return count;
+
+            if (_kafkaProducer == null)
+            {
+                _kafkaProducer = QueueHelper.CreateKafkaClient(Configuration.Connection, TransactionalId,
+                    Database.ServerStore.Server.Certificate.Certificate);
+                _kafkaProducer.InitTransactions(TimeSpan.FromSeconds(60));
+            }
+
+            void ReportHandler(DeliveryReport<string?, byte[]> report)
+            {
+                if (report.Error.IsError == false)
+                {
+                    count++;
+                }
+                else
+                {
+                    Logger.Info($"Failed to deliver message: {report.Error.Reason}");
+                    _kafkaProducer.AbortTransaction();
+                }
+            }
+
+            try
+            {
+                _kafkaProducer.BeginTransaction();
+
+                foreach (var @event in events)
+                {
+                    _kafkaProducer.Produce(@event.Topic, @event.Message, ReportHandler);
+                }
+
+                _kafkaProducer.CommitTransaction();
+            }
+            catch (Exception ex)
+            {
+                if (ex is KafkaException) _kafkaProducer.AbortTransaction();
```

Under what scenario do we want to _not_ abort the transaction here?
What happens if the abort fails? It would mask the actual error, no?

---

Agree, I will fix that.